### PR TITLE
detect/bytemath: Support out of order options

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #7565 

The intent of this PR is to support arbitrary order options for `byte_math`. During the investigation phase, it was suggested that the parser be converted to Rust for simplification. Thus, the PCRE based parser for `byte_math` was removed and replaced with a Rust based parser. @jasonish's prototype rules parser work helped guide this PR.

The C unittests were retained and many were added to the Rust based parser.

Issue: 5077

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [5077](https://redmine.openinfosecfoundation.org/issues/5077)

Describe changes:
- Converts the PCRE based parser to Rust.
- Adds unit tests to the new Rust modules
- Removes the PCRE parser from detect-bytemath.c
- Adjusts the C source modules to refer to the Rust definitions

Updates
- Address review comments from #7565 

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 8493

File: doc/userguide/rules/payload-keywords.rst [hunk 1]
```diff
@@ -431,19 +424,29 @@ Format::
...
 |		       	| - oct Converted data is represented as octal				|
 +-----------------------+-----------------------------------------------------------------------+
-| [dce]			| Allow the DCE module determine the byte order				|
+| [dce]                 | Allow the DCE module to determine the byte order                      |
 +-----------------------+-----------------------------------------------------------------------+
-| [bitmask] <value>	| The AND operator will be applied to the extracted value		|
-|			| The result will be right shifted by the number of bits equal to the	|
-|			| number of trailing zeros in the mask					|
+| [bitmask] <value>     | The AND operator will be applied to the extracted value               |
+|                       | The result will be right shifted by the number of bits equal to the   |
```
Comment (jufajardini) [inline doc/userguide/rules/payload-keywords.rst:437]: are we missing the text that was on line 444 here?

File: doc/userguide/rules/payload-keywords.rst [hunk 2]
```diff
@@ -468,11 +471,13 @@ The ``byte_jump`` keyword allows for the ability to select a ``<num of bytes>``
...
   byte_jump:<num of bytes>, <offset> [, relative][, multiplier <mult_value>] \
-	[, <endian>][, string, <num_type>][, align][, from_beginning][, from_end] \ 
+        [, <endian>][, string, <num_type>][, align][, from_beginning][, from_end] \
         [, post_offset <value>][, dce][, bitmask <value>];
 
+
+
 +-----------------------+-----------------------------------------------------------------------+
-| <num of bytes>	| The number of bytes selected from the packet to be converted		|
+| <num of bytes>       | The number of bytes selected from the packet to be converted           |
```
Comment (jufajardini) [inline doc/userguide/rules/payload-keywords.rst:480]: I think a sneaky misalignment got in!

File: doc/userguide/rules/payload-keywords.rst [hunk 3]
```diff
@@ -552,9 +558,9 @@ Format::
 |		     | - dec - Converted string represented in decimal				|
 |		     | - oct - Converted string represented in octal				|
 +--------------------+--------------------------------------------------------------------------+
-| [dce]		     | Allow the DCE module determine the byte order 				|
+| [dce]              | Allow the DCE module to determine the byte order                         |
 +--------------------+--------------------------------------------------------------------------+
-| align <align-value>| Round the extracted value up to the next 				|
+| align <align-value>| Round the extracted value up to the next                                 |
 |                    | next <align-value> byte boundary post-multiplication (if any)            |
```
Comment (jufajardini) [inline doc/userguide/rules/payload-keywords.rst:564]: Is this second `next` on purpose, or hiding in plain sight? :P

**Review by jufajardini — CHANGES_REQUESTED** [review]
Doc changes look good to me, saw just two or three sneaky things.

File: doc/userguide/rules/payload-keywords.rst (hunk 3 above)
Comment (jlucovsky) [reply doc/userguide/rules/payload-keywords.rst:564]: Good catch .. thanks.

File: doc/userguide/rules/payload-keywords.rst (hunk 1 above)
Comment (jlucovsky) [reply doc/userguide/rules/payload-keywords.rst:437]: Good catch.

**Comment by jlucovsky** [comment]
Continued in #7706

**Pull request closed without merging**